toolkit/tools/pkg/imagecustomizerlib/imagecustomizer.go

```diff
@@ -71,124 +163,209 @@ func CustomizeImageWithConfigFile(buildDir string, configFile string, imageFile
-	}
+	return allErrors
+}
 
-	// Normalize 'buildDir' path.
-	buildDirAbs, err := filepath.Abs(buildDir)
+func CustomizeImage(buildDir string, baseConfigPath string, config *imagecustomizerapi.Config, imageFile string,
+	rpmsSources []string, outputImageFile string, outputImageFormat string, outputSplitPartitionsFormat string,
+	useBaseImageRpmRepos bool, enableShrinkFilesystems bool,
+) error {
+	err := validateConfig(baseConfigPath, config, rpmsSources, useBaseImageRpmRepos)
 	if err != nil {
-		return err
+		return fmt.Errorf("invalid image config:\n%w", err)
 	}
 
-	// Create 'buildDir' directory.
-	err = os.MkdirAll(buildDirAbs, os.ModePerm)
+	imageCustomizerParameters, err := createImageCustomizerParameters(buildDir, imageFile,
+		baseConfigPath, config,
+		useBaseImageRpmRepos, rpmsSources, enableShrinkFilesystems, outputSplitPartitionsFormat,
+		outputImageFormat, outputImageFile)
 	if err != nil {
-		return err
+		return fmt.Errorf("failed to create image customizer:\n%w", err)
 	}
-
-	// Convert image file to raw format, so that a kernel loop device can be used to make changes to the image.
-	rawImageFile := filepath.Join(buildDirAbs, BaseImageName)
 	defer func() {
-		cleanupErr := file.RemoveFileIfExists(rawImageFile)
+		cleanupErr := cleanUp(imageCustomizerParameters)
 		if cleanupErr != nil {
 			if err != nil {
-				err = fmt.Errorf("%w:\nfailed to clean-up (%s): %w", err, rawImageFile, cleanupErr)
+				err = fmt.Errorf("%w:\nfailed to clean-up:\n%w", err, cleanupErr)
 			} else {
-				err = fmt.Errorf("failed to clean-up (%s): %w", rawImageFile, cleanupErr)
+				err = fmt.Errorf("failed to clean-up:\n%w", cleanupErr)
+			}
+		}
+	}()
+
+	isoBuilder, err := convertInputImageToWriteableFormat(imageCustomizerParameters)
+	if err != nil {
+		return fmt.Errorf("failed to convert input image to a raw image:\n%w", err)
+	}
+	defer func() {
+		if isoBuilder != nil {
+			cleanupErr := isoBuilder.cleanUp()
+			if cleanupErr != nil {
+				if err != nil {
+					err = fmt.Errorf("%w:\nfailed to clean-up iso builder state:\n%w", err, cleanupErr)
+				} else {
+					err = fmt.Errorf("failed to clean-up iso builder state:\n%w", cleanupErr)
+				}
 			}
 		}
 	}()
 
-	logger.Log.Infof("Creating raw base image: %s", rawImageFile)
-	err = shell.ExecuteLiveWithErr(1, "qemu-img", "convert", "-O", "raw", imageFile, rawImageFile)
+	if imageCustomizerParameters.customizeOSPartitions {
+		err = customizeOSContents(imageCustomizerParameters)
+		if err != nil {
+			return fmt.Errorf("failed to customize raw image:\n%w", err)
+		}
+	} else {
+		logger.Log.Infof("Skipping customizing OS partitions because none is specified")
+	}
+
+	err = convertWriteableFormatToOutputImage(imageCustomizerParameters, isoBuilder)
 	if err != nil {
-		return fmt.Errorf("failed to convert image file to raw format:\n%w", err)
+		return fmt.Errorf("failed to convert customized raw image to output format:\n%w", err)
+	}
+
+	logger.Log.Infof("Success!")
```

Comment:
unrelated to this PR - I always think it would be good to print the final generated image full path as part of the success message.